atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,325 @@
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            container.close();
+        });
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !closeFuture.isDone());
+        verifyNoInteractions(mockCloseable);
+
+        manuallyFinishRunWith.await();
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(closeFuture::isDone);
+    }
+
+    @Test
+    public void refreshCloseBlockedDuringRunWithInvocation()
+            throws IOException, BrokenBarrierException, InterruptedException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        CyclicBarrier runWithFutureRunsFirstBarrier = new CyclicBarrier(2);
+        CyclicBarrier manuallyFinishRunWith = new CyclicBarrier(2);
+
+        executor.execute(() -> container.executeWithResource(closeable -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            awaitOrFail(manuallyFinishRunWith);
+            verifyNoInteractions(closeable);
+        }));
+
+        Future<?> refreshableFuture = executor.submit(() -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            settableRefreshable.update(UPDATED_VALUE);
+        });
+
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !refreshableFuture.isDone());
+        verifyNoInteractions(mockCloseable);
+
+        manuallyFinishRunWith.await();
+
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(refreshableFuture::isDone);
+        verify(mockCloseable).close();
+    }
+
+    @Test
+    public void refreshUpdatesResource() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseable));
+
+        settableRefreshable.update(UPDATED_VALUE);
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseableExtra));
+    }
+
+    @Test
+    public void previousResourceClosedAfterRefresh() throws IOException {
+        // Creating a container has the side effect of managing all resources created within it.
+        getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        settableRefreshable.update(UPDATED_VALUE);
+        verify(mockCloseable).close();
+    }
+
+    @Test
+    public void latestResourceClosedAfterClose() throws IOException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        settableRefreshable.update(UPDATED_VALUE);
+        container.close();
+
+        verify(mockCloseableExtra).close();
+    }
+
+    @Test
+    public void noFurtherResourceCreatedAfterClose() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        container.close();
+        settableRefreshable.update(UPDATED_VALUE);
+
+        verify(closeableFactory, times(1)).apply(any());
```

Comment:
`any()` is best used in the negative, i.e.

`verify(myObject, never()).myMethod(any())`

If verifying in the positive direction, it's generally better to be more precise - you can confirm the _exact_ invocations that way, and it makes it clearer what you're doing.

Reply:
also, fwiw, `verify(...)` defaults to `times(1)`